File: OC-Demos/TestRuntime/TestRuntime/ViewController.m
//
//  ViewController.m
//  TestRuntime
//
//  Created by wei on 2016/12/26.
//  Copyright © 2016年 wei. All rights reserved.
//

#import "ViewController.h"

@interface ViewController ()

@property (strong, nonatomic) NSMutableArray *titles;
@property (strong, nonatomic) NSMutableArray *classStrs;

@end

static NSString *cellIdentity = @"cell";

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    
    self.title = @"Runtime Test Demo";
    self.titles = @[].mutableCopy;
    self.classStrs  = @[].mutableCopy;
    
    [self.tableView registerClass:[UITableViewCell class] forCellReuseIdentifier:cellIdentity];
    
    
    [self addClassStr:@"ClassAndObjectVC" title:@"Class&Object Demo"];
    [self addClassStr:@"AddProperty_Ivar_MethodDynamicVC" title:@"AddProperty&Ivar&MethodDynamic Demo"];
    [self addClassStr:@"AddAssociatedOjectVC" title:@"AddAssociatedOject Demo"];
    [self addClassStr:@"AboutMethodVC" title:@"AboutMethod&消息转发(message forwarding) Demo"];
    [self addClassStr:@"MethodSwizzlingVC" title:@"MethodSwizzling Demo"];
    [self addClassStr:@"MethodSwizzlingVC2" title:@"MethodSwizzling Demo2"];
    
}

- (void)addClassStr:(NSString *)classStr title:(NSString *)title{
    [self.titles addObject:title];
    [self.classStrs addObject:classStr];
}


#pragma mark -UITableViewDelegate
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section{
    return self.titles.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath{
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:cellIdentity forIndexPath:indexPath];
    cell.textLabel.text = self.titles[indexPath.row];
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath{
    Class class = NSClassFromString(self.classStrs[indexPath.row]);
    
    if (class) {
        UIViewController *vc = [class new];
        vc.view.backgroundColor = [UIColor whiteColor];
        [self.navigationController pushViewController:vc animated:YES];
        vc.title = self.titles[indexPath.row];
        
        UILabel *label = [[UILabel alloc] initWithFrame:CGRectMake(0, 0, [UIScreen mainScreen].bounds.size.width, 10)];
        label.textAlignment = NSTextAlignmentCenter;
        label.backgroundColor = [UIColor clearColor];
        [vc.view addSubview:label];
        label.text = [NSString stringWithFormat:@"see code in Class %@\n\nsee rusult in console", NSStringFromClass(class)];
        label.numberOfLines = 0;
        [label sizeToFit];
        label.center = vc.view.center;
    }
    
    [self.tableView deselectRowAtIndexPath:indexPath animated:NO];
}

- (CGFloat)tableView:(UITableView *)tableView heightForFooterInSection:(NSInteger)section{
    return 0.01;
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


@end

File: Swift-Demos/TestSwift/TestSwift/ViewController.swift
//
//  ViewController.swift
//  TestSwift
//
//  Created by wei on 2017/3/22.
//  Copyright © 2017年 wei. All rights reserved.
//

import UIKit

class ViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        
        let my = myClass(name:"name", age:67)
        my.desc()
        
        let square = Square(sideLength: 65)
        square.desc()
        
        let label = UILabel(frame: CGRect(x:100, y:100, width:200, height:300))
        label.backgroundColor = UIColor.red
        self.view.addSubview(label)
        
        var array = [String]()
        array.append("nnnnnn")
        array.append(contentsOf: ["点击放大就返回", "fjdfdjkf", "893dfuj"])
        
        for (key, value) in array.enumerated() {
            print(key, value)
        }
        
        var array2 = ["ghfhgfghfhgfgh", "678", "454", "99889"]
        print(array2)
        var reverseArray2 = array2.sorted(by:{(s1:String, s2:String)->Bool in
            return s1>s2
        })
        print(reverseArray2)
        
        var reverseArray = array.sorted(by: <)
        print(reverseArray)
        
        let count = array2.count
        switch count {
        case 3:
            print("count > 3")
            fallthrough
        case 2:
            print("count == 2")
        default:
            print("default")
        }
        
        var a = 10, b = 20
        print("a= \(a), b= \(b)")
        switchNum(&a, &b)
        print("a= \(a), b= \(b)")
        
        
        var numbers = [23, 43, 5, 45, 234];
        print(numbers.sorted(by:>))
        
    }
    
    func switchNum(_ a : inout Int,_ b : inout Int) {
        let temp = a
        a = b
        b = temp
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }


}

